### PR TITLE
Feature/psp tests smock

### PR DESCRIPTION
**Pull request by novaknole**
## Description

Please include a summary of the change and be sure you follow the contributions rules we do provide [here](./CONTRIBUTIONS.md)

Task: [ID]()

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] I have selected the correct base branch.
- [ ] I have performed a self-review of my own code.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.
- [ ] I ran all tests with success and extended them if necessary.
- [ ] I have updated the ``CHANGELOG.md`` file in the root folder.
- [ ] I have tested my code on the test network.

### CODE REVIEW

File: packages/contracts/test/framework/plugin/plugin-setup-processor.ts [hunk 1]
```diff
@@ -427,135 +540,232 @@ describe.skip('Plugin Setup Processor', function () {
...
+              pluginRepoPointer,
+              firstPreparedPermissions,
+              firstPreparedHelpers
+            )
           )
-        )
-          .to.emit(psp, EVENTS.InstallationApplied)
-          .withArgs(targetDao.address, plugin);
-      });
+        ).to.be.reverted;
```
Comment (mathewmeconry) [inline packages/contracts/test/framework/plugin/plugin-setup-processor.ts:754]: add at least one non-static call and check for the revert `PluginAlreadyInstalled` so that we can verify we go into the right code path.

File: packages/contracts/test/framework/plugin/plugin-setup-processor.ts [hunk 2]
```diff
@@ -577,157 +789,253 @@ describe.skip('Plugin Setup Processor', function () {
...
-            Object.values(perm)
+        await expect(
+          psp.prepareUninstallation(
+            targetDao.address,
+            createPrepareUninstallationParams(
+              proxy,
+              pluginRepoPointer,
+              helpersUV1,
+              EMPTY_DATA
+            )
```
Comment (mathewmeconry) [inline packages/contracts/test/framework/plugin/plugin-setup-processor.ts:941]: check that it reverts with `InvalidAppliedSetupId` so we know that we landed in the right code path

File: packages/contracts/test/framework/plugin/plugin-setup-processor.ts [hunk 3]
```diff
@@ -878,340 +1238,676 @@ describe.skip('Plugin Setup Processor', function () {
...
+        ],
+        secondPreparationInitData,
+        secondPreparationPermissions,
+        secondPreparationHelpers
+      );
 
-        await expect(
-          psp.applyUpdate(
-            targetDao.address,
+      // confirm that now preparations can't be applied anymore
```
Comment (mathewmeconry) [inline packages/contracts/test/framework/plugin/plugin-setup-processor.ts:1844]: check at least one of them as non-static so we can make sure we land in the rigth code path

File: packages/contracts/test/framework/plugin/plugin-setup-processor.ts [hunk 4]
```diff
@@ -1769,244 +2293,102 @@ describe.skip('Plugin Setup Processor', function () {
...
 
   // If the base contracts don't change from current and new plugin setups,
   // PluginSetupProcessor shouldn't call `upgradeTo` or `upgradeToAndCall`
   // on the plugin. The below check for this still is not 100% ensuring,
   // As function `upgradeTo` might be called but event `Upgraded`
   // not thrown(OZ changed the logic or name) which will trick the test to pass..
-  const currentImpl = await currentVersionSetup.getImplementationAddress();
-  const newImpl = await newVersionSetup.getImplementationAddress();
+  const currentImpl = await currentPluginSetup.getImplementationAddress();
+  const newImpl = await newPluginSetup.getImplementationAddress();
```
Comment (mathewmeconry) [inline packages/contracts/test/framework/plugin/plugin-setup-processor.ts:2360]: Giorgi make working?

**Review by mathewmeconry — CHANGES_REQUESTED** [review]
Good tests mate!
I would love to also see the following path in the tests:
Install -> Uninstall -> Install -> Update

Just to make sure that not only the Uninstall -> prepareInstallation works but also applying said preparation

File: packages/contracts/test/framework/plugin/plugin-setup-processor.ts (hunk 3 above)
Comment (novaknole) [reply packages/contracts/test/framework/plugin/plugin-setup-processor.ts:1844]: Removed `callStatic` check ! it was not needed.

File: packages/contracts/test/framework/plugin/plugin-setup-processor.ts (hunk 1 above)
Comment (novaknole) [reply packages/contracts/test/framework/plugin/plugin-setup-processor.ts:754]: yep, removed staticcall.

File: packages/contracts/src/test/plugin/Clonable/PluginCloneableMock.sol [hunk 5]
```diff
@@ -14,7 +14,16 @@ contract PluginCloneableV1Mock is PluginCloneable {
     }
 }
 
-contract PluginCloneableV2Mock is PluginCloneable {
+// Doesn't support IPlugin Interface.
+contract PluginCloneableV2Mock {
```
Comment (Michael-A-Heuer) [inline packages/contracts/src/test/plugin/Clonable/PluginCloneableMock.sol:18]: I would name the new one `WrongPluginCloneableV1Mock` or something similar. 

The idea behind V1->V2 was to show how a storage variable can be added.
V1->InvalidV2->V3 makes it confusing.

File: packages/contracts/src/test/plugin/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol [hunk 6]
```diff
@@ -42,10 +42,11 @@ contract PluginUUPSUpgradeableSetupV1Mock is PluginSetup {
 }
 
 contract PluginUUPSUpgradeableSetupV1MockBad is PluginUUPSUpgradeableSetupV1Mock {
+    
     function prepareInstallation(
         address _dao,
         bytes memory
-    ) public pure override returns (address plugin, PreparedSetupData memory preparedSetupData) {
+    ) public view override returns (address plugin, PreparedSetupData memory preparedSetupData) {
```
Comment (Michael-A-Heuer) [inline packages/contracts/src/test/plugin/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol:49]: This adds a compiler warning.

```
Warning: Unused function parameter. Remove or comment out the variable name to silence this warning.
  --> src/test/plugin/Clonable/PluginCloneableMock.sol:21:25:
   |
21 |     function initialize(IDAO _dao) external {
   |                         ^^^^^^^^^


Warning: Function state mutability can be restricted to pure
  --> src/test/plugin/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol:46:5:
   |
46 |     function prepareInstallation(
   |     ^ (Relevant source part starts here and spans across multiple lines).
```
Please revert and fix the other compiler warning as well.

File: packages/contracts/test/test-utils/psp/atomic-helpers.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  applyUpdate,
+  prepareInstallation,
+  prepareUninstallation,
+  prepareUpdate,
+} from './wrappers';
+import {UPGRADE_PERMISSIONS} from '../permissions';
+
+const {UPGRADE_PLUGIN_PERMISSION_ID} = UPGRADE_PERMISSIONS;
+
+// TODO: put in common or something.
```
Comment (Michael-A-Heuer) [inline packages/contracts/test/test-utils/psp/atomic-helpers.ts:16]: Forgotten TODO

File: packages/contracts/test/framework/plugin/plugin-setup-processor.ts [hunk 8]
```diff
@@ -14,32 +15,73 @@ import {
...
   PluginRepo,
   DAO,
+  PluginUUPSUpgradeableSetupV2Mock__factory,
+  PluginUUPSUpgradeableSetupV1MockBad__factory,
+  PluginUUPSUpgradeableSetupV4Mock__factory,
+  PluginCloneableSetupV2Mock__factory,
+  PluginCloneableSetupV3Mock__factory,
+  PluginUUPSUpgradeableSetupV1Mock__factory,
+  PluginCloneableSetupV1Mock__factory,
+  IPluginSetup__factory,
```
Comment (Michael-A-Heuer) [inline packages/contracts/test/framework/plugin/plugin-setup-processor.ts:30]: unused import

File: packages/contracts/test/test-utils/plugin-setup-processor.ts [hunk 9]
```diff
@@ -1,9 +1,9 @@
 import {ethers} from 'hardhat';
-import {findEvent} from '../../utils/event';
+import {utils, constants} from 'ethers';
```
Comment (Michael-A-Heuer) [inline packages/contracts/test/test-utils/plugin-setup-processor.ts:2]: Remove unused imports

File: packages/contracts/test/test-utils/psp/wrappers.ts [hunk 10]
```diff
@@ -0,0 +1,139 @@
+import {BytesLike} from 'ethers';
+import {PluginRepo, PluginSetupProcessor} from '../../../typechain';
```
Comment (Michael-A-Heuer) [inline packages/contracts/test/test-utils/psp/wrappers.ts:2]: PluginRepo is not used

File: packages/contracts/test/framework/plugin/plugin-setup-processor.ts [hunk 11]
```diff
@@ -1769,236 +2309,93 @@ describe.skip('Plugin Setup Processor', function () {
...
   // PluginSetupProcessor shouldn't call `upgradeTo` or `upgradeToAndCall`
   // on the plugin. The below check for this still is not 100% ensuring,
   // As function `upgradeTo` might be called but event `Upgraded`
   // not thrown(OZ changed the logic or name) which will trick the test to pass..
-  const currentImpl = await currentVersionSetup.getImplementationAddress();
-  const newImpl = await newVersionSetup.getImplementationAddress();
+  const currentImpl = await currentPluginSetup.getImplementationAddress();
+  const newImpl = await newPluginSetup.getImplementationAddress();
+
+  // TODO: Giorgi Not working..
```
Comment (Michael-A-Heuer) [inline packages/contracts/test/framework/plugin/plugin-setup-processor.ts:2378]: Better remove the comments and push them to a separate branch and create a follow-up task.

File: packages/contracts/src/test/plugin/Clonable/PluginCloneableMock.sol (hunk 5 above)
Comment (novaknole) [reply packages/contracts/src/test/plugin/Clonable/PluginCloneableMock.sol]: https://github.com/aragon/core/pull/242/commits/48b211bbbdf5d20660a929aa53fb37e89ef5c3e9 (resolve if you agree)

File: packages/contracts/src/test/plugin/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol (hunk 6 above)
Comment (novaknole) [reply packages/contracts/src/test/plugin/UUPSUpgradeable/PluginUUPSUpgradeableSetupMock.sol]: https://github.com/aragon/core/pull/242/commits/48b211bbbdf5d20660a929aa53fb37e89ef5c3e9 (resolve if you agree)

File: packages/contracts/test/test-utils/psp/atomic-helpers.ts (hunk 7 above)
Comment (novaknole) [reply packages/contracts/test/test-utils/psp/atomic-helpers.ts]: https://github.com/aragon/core/pull/242/commits/ba117edd405db7bb20ee3aa8870d1e4b57d7baab

File: packages/contracts/test/framework/plugin/plugin-setup-processor.ts [hunk 12]
```diff
@@ -260,98 +322,151 @@ describe.skip('Plugin Setup Processor', function () {
...
-        ({
-          plugin: plugin,
-          helpers: helpersCV1,
-          permissions: permissionsCV1,
-        } = await prepareInstallation(
+      // 1. prepareInstall for pluginId1 => setupId1
+      // 2. applyInstall for pluginId1 => setupId1
+      // 3. uninstall the plugin with applyUninstall.
+      // 4. prepareInstall for pluginId1 => setupId1 which succeeds.
+      it('EDGE-CASE: allows to prepare plugin installation with the same address and setupId if it was installed and then uninstalled', async () => {
```
Comment (Michael-A-Heuer) [inline packages/contracts/test/framework/plugin/plugin-setup-processor.ts:394]: Isn't this a "Reinstallation"? Think if the test naming can be improved
```suggestion
      it('reinstalls a plugin contract (same address and `prepareSetupId` that has been uinstalled previously', async () => {
```

File: packages/contracts/test/test-utils/plugin-setup-processor.ts (hunk 9 above)
Comment (novaknole) [reply packages/contracts/test/test-utils/plugin-setup-processor.ts]: https://github.com/aragon/core/pull/242/commits/74729dae32addd7008613ddca878730aa50ed628

File: packages/contracts/test/framework/plugin/plugin-setup-processor.ts (hunk 11 above)
Comment (novaknole) [reply packages/contracts/test/framework/plugin/plugin-setup-processor.ts]: https://github.com/aragon/core/pull/242/commits/74729dae32addd7008613ddca878730aa50ed628

File: packages/contracts/test/framework/plugin/plugin-setup-processor.ts [hunk 13]
```diff
@@ -427,135 +538,233 @@ describe.skip('Plugin Setup Processor', function () {
...
-        );
-        const {plugin, permissions: permissions} = await prepareInstallation(
-          psp,
-          targetDao.address,
-          setupUV1Bad.address,
-          repoU.address,
-          dataUser1
-        );
+        // really don't matter what we choose here for the plugin address.
+        const pluginAddress = ownerAddress;
```
Comment (Michael-A-Heuer) [inline packages/contracts/test/framework/plugin/plugin-setup-processor.ts:545]:
```suggestion
        const wrongPluginAddress = ownerAddress;
```

File: packages/contracts/test/framework/plugin/plugin-setup-processor.ts [hunk 14]
```diff
@@ -427,135 +538,233 @@ describe.skip('Plugin Setup Processor', function () {
...
 
-      it('reverts if the installation was not prepared', async () => {
-        let unpreparedPlugin;
-        let unpreparedHelpers;
-        let unPreparedPermissions;
+      // 1. call prepareinstall 2 times for the same plugin version
+      // to get 2 preparations with same plugin address, but different setup ids.
+      // 2. call applyInstall for one of them and see that 2nd one
+      // would no longer be valid for the installation even though it was valid before.
+      it('EDGE-CASE: reverts for all preparation if one of them was already applied for the install', async () => {
```
Comment (Michael-A-Heuer) [inline packages/contracts/test/framework/plugin/plugin-setup-processor.ts:641]: grammar

**Review by Michael-A-Heuer — APPROVED** [review]

**Review by brickpop — APPROVED** [review]
